### ISSUE TITLE
Retrieving raw predictions in c# wrapper

### ISSUE THREAD

**Issue by Wirack** [JohnLangford/vowpal_wabbit#1118]
I am currently working with oaa, and can expose scores with use of logistic loss.
However, when I use another loss function I am not able to expose the scores.

I found that --raw_predictions can be used to extract raw predictions in vowpal wabbit.

How can I extract raw predictions in the c# wrapper, apart from outputting to a file?

And if that is not available currently, where should I modify the wrapper (and maybe vw?) to be able to expose those raw predictions?

**Comment by eisber** [comment]
You should modify the prediction factory, but I suppose we need some advice from John on how we should report those back through the example data structure. I don’t think it’s include yet.

**Comment by Wirack** [comment]
I have begun implementing that, but stumbled into the same problem with raw predictions not being exposed in the example data structure.

I am wondering if oaa.cc should be adjusted to expose this in the example data structure, or if it is possible to implement something more general?

**Comment by eisber** [comment]
Without taking a too close look, we could introduce a commonly used field used by each instance.

Markus

**Comment by Wirack** [comment]
Yes, I think that is the right approach, and is also discussed [here](https://github.com/JohnLangford/vowpal_wabbit/issues/918).
I also think the '--scores' flag would be  a good way to go.

**Comment by mttdbrd** [comment]
I wrote a method in the Java wrapper which shows how to extract raw predictions without writing to a file. You can see it here: [https://github.com/JohnLangford/vowpal_wabbit/compare/master...getservice:master](url)

**Comment by Wirack** [comment]
Thanks, will take a look when I have some free time.

**Comment by JohnLangford** [comment]
You can now pass --scores to oaa.cc and it will expose all predictions via pred.scalars.

**Comment by Wirack** [comment]
Thanks, that's a nice clean solution, 
And since VowpalWabbitScalarsPredictionFactory already exposes pred.scalars I would say this issue is resolved.

**Comment by Wirack** [comment]
It seems VowpalWabbitScalarsPredictionFactory does not allow for multiclass scores, but VowpalWabbitPredictionType.MultiClassProbabilities can be used instead.

**Comment by marcocaccin** [comment]
I'm using this feature with `hinge` loss and I'm noticing that all scores are capped in the [-1, 1] range. Is this intended, and what's the reason?
(I am intuitively assuming that a score with hinge loss is the margin of a linear SVM, correct me if I'm wrong!)

**Comment by Wirack** [comment]
I stumbled into a similar issue, described [here](https://github.com/JohnLangford/vowpal_wabbit/issues/1143).
It is related to this [code](https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/gd.cc#L309-L312) which adjusts the scores. In my fork of the repo I have removed the stabilization that occurs there.

As described [here](https://github.com/JohnLangford/vowpal_wabbit/issues/1143) it is due to the following:

From John Langford:
> It’s not underflow/overflow, but rather a stabilization of regression to the range of observed values.

**Comment by marcocaccin** [comment]
Ah, I see! Thank you for the clarification, will use your fork then :)

**Comment by JohnLangford** [comment]
Note that there are flags to adjust the scores to work over any desired
range.

-John

**Comment by Wirack** [comment]
Oh I see! I will use that instead then :)
`--min_prediction arg           Smallest prediction to output`
`--max_prediction arg           Largest prediction to output`

**Comment by marcocaccin** [comment]
...and it works out of the box, thanks for the heads up! I read the documentation for those parameters but didn't connect the dots.
 
Just for reference for whoever lands here in the future, if you want to get prediction scores the same way you would use scikit-learn's `predict_proba()` or `decision_function()`:
- Train: 
`vw train.data [whatever] -f model.vw --scores --min_prediction -$margin --max_prediction $margin`
- Predict scores: 
`vw test.data -t -i model.vw -r preds.txt`

where `margin` depends on which loss function is used: for example, `10` is more than enough for of `hinge` loss when features have order of magnitude 1 (e.g., normalized and/or when dealing with text data).
